src/syncback/mod.rs

```diff
@@ -0,0 +1,495 @@
+        // We can quickly check that two subtrees are identical and if they are,
+        // skip reconciling them.
+        if let Some(old_ref) = snapshot.old {
+            match (old_hashes.get(&old_ref), new_hashes.get(&snapshot.new)) {
+                (Some(old), Some(new)) => {
+                    if old == new {
+                        log::trace!(
+                            "Skipping {inst_path} due to it being identically hashed as {old:?}"
+                        );
+                        continue;
+                    }
+                }
+                _ => unreachable!("All Instances in both DOMs should have hashes"),
+            }
+        }
+
+        if !is_valid_path(&ignore_patterns, project_path, &snapshot.path) {
+            log::debug!("Skipping {inst_path} because its path matches ignore pattern");
+            continue;
+        }
+        if let Some(syncback_rules) = &project.syncback_rules {
+            // Ignore trees;
+            for ignored in &syncback_rules.ignore_trees {
+                if inst_path.starts_with(ignored.as_str()) {
+                    log::debug!("Tree {inst_path} is blocked by project");
+                    continue 'syncback;
+                }
+            }
+        }
+
+        let middleware = get_best_middleware(&snapshot);
+
+        log::trace!(
+            "Middleware for {inst_path} is {:?} (path is {})",
+            middleware,
+            snapshot.path.display()
+        );
+
+        if matches!(middleware, Middleware::Json | Middleware::Toml) {
+            log::warn!("Cannot syncback {middleware:?} at {inst_path}, skipping");
+            continue;
+        }
+
+        let mut syncback_res = middleware.syncback(&snapshot);
+        if syncback_res.is_err() && middleware == Middleware::Dir {
+            let new_middleware = match env::var(DEBUG_MODEL_FORMAT_VAR) {
+                Ok(value) if value == "1" => Middleware::Rbxmx,
+                Ok(value) if value == "2" => Middleware::JsonModel,
+                _ => Middleware::Rbxm,
+            };
+
+            let file_name = snapshot
+                .path
+                .file_name()
+                .and_then(|s| s.to_str())
+                .context("Directory middleware should have a name in its path")?;
+            let mut path = snapshot.path.clone();
+            path.set_file_name(format!(
+                "{file_name}.{}",
+                extension_for_middleware(new_middleware)
+            ));
+
+            let new_snapshot = snapshot.with_new_path(path, snapshot.new, snapshot.old);
+            log::warn!(
+                "Could not syncback {inst_path} as a Directory, it will \
+                instead be synced back as a {new_middleware:?}."
+            );
+            syncback_res = new_middleware.syncback(&new_snapshot);
+        }
+        let syncback = syncback_res.with_context(|| format!("Failed to syncback {inst_path}"))?;
```

Comment:
This code doesn't tell users why an instance was not able to be synced back to a directory. We might also consider using pattern matching here rather than a mutable variable, to make the code a little clearer:
```suggestion
        let syncback = match middleware.syncback(&snapshot) {
            Ok(syncback) => syncback,
            Err(err) if middleware == Middleware::Dir => {
                let new_middleware = match env::var(DEBUG_MODEL_FORMAT_VAR) {
                    Ok(value) if value == "1" => Middleware::Rbxmx,
                    Ok(value) if value == "2" => Middleware::JsonModel,
                    _ => Middleware::Rbxm,
                };

                let file_name = snapshot
                    .path
                    .file_name()
                    .and_then(|s| s.to_str())
                    .context("Directory middleware should have a name in its path")?;
                let mut path = snapshot.path.clone();
                path.set_file_name(format!(
                    "{file_name}.{}",
                    extension_for_middleware(new_middleware)
                ));

                let new_snapshot = snapshot.with_new_path(path, snapshot.new, snapshot.old);
                log::warn!(
                    "Could not syncback {inst_path} as a Directory because {err}
It will instead be synced back as a {new_middleware:?}."
                );
                new_middleware
                    .syncback(&new_snapshot)
                    .with_context(|| "Failed to syncback {inst_path}")?
            }
            Err(err) => anyhow::bail!("Failed to syncback {inst_path} because {err}"),
        };
```

(I'm also noticing that attempting to sync back e.g. a folder containing children with duplicate names causes this log to be printed twice - any idea why?)

Reply:
I'm not sure why the duplication is happening, but I'll take a looksie.